File: seqkit/sample.go
package seqkit

import (
	"fmt"
	"ignis/driver/api"
)

type SeqKitSampleOptions struct {
	inner SampleOptions
}

type SampleOptions struct {
	Config     KitConfig
	Seed       *int
	Number     *int
	Proportion *float32
	TwoPass    *bool
}

func (this *SampleOptions) setDefaults() *SampleOptions {
	this.Config.setDefaults()
	setDefault(&this.Seed, 11)
	setDefault(&this.Number, 0)
	setDefault(&this.Proportion, 0)
	setDefault(&this.TwoPass, false)

	return this
}

func (this *SeqKitSampleOptions) Config(v *SeqKitConfig) *SeqKitSampleOptions {
	this.inner.Config = v.inner
	return this
}

func (this *SeqKitSampleOptions) Seed(v int) *SeqKitSampleOptions {
	this.inner.Seed = &v
	return this
}

func (this *SeqKitSampleOptions) Number(v int) *SeqKitSampleOptions {
	this.inner.Number = &v
	return this
}

func (this *SeqKitSampleOptions) Proportion(v float32) *SeqKitSampleOptions {
	this.inner.Proportion = &v
	return this
}

func (this *SeqKitSampleOptions) TwoPass(v bool) *SeqKitSampleOptions {
	this.inner.TwoPass = &v
	return this
}

func Sample(input *api.IDataFrame[string], o *SeqKitSampleOptions) (*api.IDataFrame[string], error) {
	if o == nil {
		o = &SeqKitSampleOptions{}
	}
	opts := o.inner
	opts.setDefaults()

	if *opts.Number == 0 && *opts.Proportion == 0 {
		return nil, fmt.Errorf("one of flags -n (--number) and -p (--proportion) needed")
	}

	if *opts.Number < 0 {
		return nil, fmt.Errorf("value of -n (--number) and should be greater than 0")
	}
	if *opts.Proportion < 0 || *opts.Proportion > 1 {
		return nil, fmt.Errorf("value of -p (--proportion) (%f) should be in range of (0, 1]", *opts.Proportion)
	}

	fraction := float64(*opts.Proportion)
	if *opts.Number > 0 {
		if !*opts.TwoPass {
			if err := input.Cache(api.PRESERVE); err != nil {
				return nil, err
			}
		}
		n, err := input.Count()
		if err != nil {
			return nil, err
		}
		fraction = float64(*opts.Number) / float64(n)
	}

	return input.Sample(false, fraction, *opts.Seed)
}

File: bigseqkit-py/bigseqkit/sample.py
from bigseqkit.helper import _setDefault, _libSource, _config, _optionsToString, _parseKargs, SeqKitConfig, IDataFrame


class SeqKitSampleOptions:

    def __init__(self):
        self.__inner = SampleOptions()

    def config(self, v: SeqKitConfig):
        self.__inner.Config = _config(v)

    def seed(self, v: int):
        self.__inner.Seed = v

    def number(self, v: int):
        self.__inner.Number = v

    def proportion(self, v: float):
        self.__inner.Proportion = v

    def _run(self, input: IDataFrame, **kwargs):
        opts = self.__inner
        _parseKargs(opts, kwargs)
        opts.setDefaults()

        if opts.Number == 0 and opts.Proportion == 0:
            raise RuntimeError("one of flags -n (--number) and -p (--proportion) needed")

        if opts.Number < 0:
            raise RuntimeError("value of -n (--number) and should be greater than 0")

        if opts.Proportion < 0 or opts.Proportion > 1:
            raise RuntimeError(f"value of -p (--proportion) ({str(opts.Proportion)}) should be in range of (0, 1]")

        fraction = opts.Proportion
        if opts.Number > 0:
            n = input.count()
            fraction = opts.Number / n

        return input.sample(False, fraction, opts.Seed)

class SampleOptions:

    def __init__(self):
        self.Config = None  # KitConfig
        self.Seed = None  # int
        self.Number = None  # int
        self.Proportion = None  # float

    def setDefaults(self):
        _setDefault(self, "Config", _config(SeqKitConfig())).setDefaults()
        _setDefault(self, "Seed", 11)
        _setDefault(self, "Number", 0)
        _setDefault(self, "Proportion", 0)


def sample(input: IDataFrame, o: SeqKitSampleOptions = None, **kwargs):
    if o is None:
        o = SeqKitSampleOptions()
    return o._run(input, **kwargs)
